velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -68,20 +69,158 @@ void WindowFuzzer::addWindowFunctionSignatures(
+
+  return std::make_tuple(windowType, frameStartBoundType, frameEndBoundType);
+}
+
+// For frames with k RANGE PRECEDING/FOLLOWING, Velox requires the application
+// to add columns with the range frame boundary value computed according to the
+// frame type.
+// If the frame is k PRECEDING :
+// frame_boundary_value = current_order_by - k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by + k (for descending ORDER BY)
+// If the frame is k FOLLOWING :
+// frame_boundary_value = current_order_by + k (for ascending ORDER BY)
+// frame_boundary_value = current_order_by - k (for descending ORDER BY)
+template <typename T>
+T WindowFuzzer::genOffsetAtIdx(
+    T* offsetCol,
+    vector_size_t idx,
+    T offsetValue,
+    core::WindowNode::BoundType frameBoundType,
+    WindowFuzzer::SortingKeyAndOrder& orderByKey) {
+  auto val = offsetCol[idx];
+  if ((frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       orderByKey.sortOrder_.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       !orderByKey.sortOrder_.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetCol[idx] - offsetValue;
+    } else {
+      return checkedMinus<T>(offsetCol[idx], offsetValue);
     }
-  };
-  auto isRowsFrame =
-      boost::random::uniform_int_distribution<uint32_t>(0, 1)(rng_);
-  auto frameTypeString = frameType(isRowsFrame);
+  }
+
+  if ((frameBoundType == core::WindowNode::BoundType::kFollowing &&
+       orderByKey.sortOrder_.isAscending()) ||
+      (frameBoundType == core::WindowNode::BoundType::kPreceding &&
+       !orderByKey.sortOrder_.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetCol[idx] + offsetValue;
+    } else {
+      return checkedPlus<T>(offsetCol[idx], offsetValue);
+    }
+  }
+
+  VELOX_UNREACHABLE(
+      "sortOrder ascending {} frameBoundType {}",
+      orderByKey.sortOrder_.toString(),
+      core::WindowNode::boundTypeName(frameBoundType));
+}
+
+template <typename T>
+std::optional<std::string> WindowFuzzer::addKRangeFrameColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    std::string& columnName,
+    SortingKeyAndOrder& orderByKey) {
+  T offsetValue;
+  if constexpr (!(std::is_same_v<T, bool> || std::is_same_v<T, StringView> ||
+                  std::is_same_v<T, Timestamp> ||
+                  std::is_same_v<T, UnknownValue>)) {
+    auto size = vectorFuzzer_.getOptions().vectorSize;
+    velox::test::VectorMaker vectorMaker{pool_.get()};
+
+    auto type = CppToType<T>::create();
+    VectorPtr fuzzOffset = vectorFuzzer_.fuzzConstant(type, 1);
+    offsetValue = fuzzOffset->as<ConstantVector<T>>()->valueAt(0);
+
+    while (offsetValue < 0) {
+      fuzzOffset = vectorFuzzer_.fuzzConstant(type, 1);
+      offsetValue = fuzzOffset->as<ConstantVector<T>>()->valueAt(0);
+    }
```

Comment:
Instead of using a while loop, let's determine the random offset in the valid range first and then create a constant vector from it.

Reply:
Reverted the while loop, this was added only for debugging purpose.